drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2434 @@
+	uint32_t i3c_clock = 0;
+	uint32_t i2c_bus_freq = data->drv_data.ctrl_config.scl.i2c;
+	uint32_t i3c_bus_freq = data->drv_data.ctrl_config.scl.i3c;
+
+	if (clock_control_get_rate(clk, (clock_control_subsys_t)&cfg->pclken[0], &i3c_clock) < 0) {
+		LOG_ERR("Failed call clock_control_get_rate(pclken[0])");
+		return -EIO;
+	}
+
+	uint8_t scll_od = 0;
+	uint8_t sclh_i2c = 0;
+	uint8_t scll_pp = 0;
+	uint8_t sclh_i3c = 0;
+	uint32_t clk_wave = 0;
+	int ret;
+
+	LOG_DBG("I3C Clock = %u, I2C Bus Freq = %u, I3C Bus Freq = %u", i3c_clock, i2c_bus_freq,
+		i3c_bus_freq);
+
+	ret = i3c_stm32_calc_scll_od_sclh_i2c(i2c_bus_freq, i3c_clock, &scll_od, &sclh_i2c);
+	if (ret != 0) {
+		LOG_ERR("Cannot calculate the timing for TimingReg0, err=%d", ret);
+		return ret;
+	}
+
+	ret = i3c_stm32_calc_scll_pp_sclh_i3c(i3c_bus_freq, i3c_clock, &scll_pp, &sclh_i3c);
+	if (ret != 0) {
+		LOG_ERR("Cannot calculate the timing for TimingReg0, err=%d", ret);
+		return ret;
+	}
+
+	clk_wave = ((uint32_t)sclh_i2c << 24) | ((uint32_t)scll_od << 16) |
+		   ((uint32_t)sclh_i3c << 8) | (scll_pp);
+
+	LOG_DBG("TimigReg0 = 0x%08x", clk_wave);
+
+	LL_I3C_ConfigClockWaveForm(i3c, clk_wave);
+
+	return 0;
+}
+/**
+ * @brief Get current configuration of the I3C hardware.
+ *
+ * @param[in] dev Pointer to controller device driver instance.
+ * @param[in] type Type of configuration.
+ * @param[in,out] config Pointer to the configuration parameters.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO General Input/Output errors.
+ * @retval -ENOSYS If not implemented.
+ */
+static int i3c_stm32_config_get(const struct device *dev, enum i3c_config_type type, void *config)
+{
+	struct i3c_stm32_data *data = dev->data;
+
+	if ((type != I3C_CONFIG_CONTROLLER) || (config == NULL)) {
+		return -EINVAL;
+	}
+
+	(void)memcpy(config, &data->drv_data.ctrl_config, sizeof(data->drv_data.ctrl_config));
+
+	return 0;
+}
+
+static int i3c_stm32_config_ctrl_bus_char(const struct device *dev)
+{
+	const struct i3c_stm32_config *cfg = dev->config;
+	struct i3c_stm32_data *data = dev->data;
+	const struct device *clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+	I3C_TypeDef *i3c = cfg->i3c;
+	uint32_t i3c_clock = 0;
+	uint32_t i2c_bus_freq = data->drv_data.ctrl_config.scl.i2c;
+
+	uint8_t free = 0;
+	uint8_t aval = 0;
+
+	if (clock_control_get_rate(clk, (clock_control_subsys_t)&cfg->pclken[0], &i3c_clock) < 0) {
+		LOG_ERR("Failed call clock_control_get_rate(pclken[0])");
+		return -EIO;
+	}
+
+	/* Satisfying I3C start timing min timing will statisfy the rest of the conditions */
+	if (i2c_bus_freq > 400000) {
+		/* Mixed bus with I2C FM+ device */
+		free = (uint8_t)ceil((STM32_I3C_TBUF_FMP_MIN_NS * i3c_clock / 1e9 - 0.5) / 2);
+	} else if (i2c_bus_freq > 0) {
+		/* Mixed bus with I2C FM device */
+		free = (uint8_t)ceil((STM32_I3C_TBUF_FM_MIN_NS * i3c_clock / 1e9 - 0.5) / 2);
+	} else {
+		/* Pure I3C bus */
+		free = (uint8_t)ceil((STM32_I3C_TCAS_MIN_NS * i3c_clock / 1e9 - 0.5) / 2);
+	}
```

Comment:
I don't think this is best way to determine if it should be in pure, mixed fast, or mixed slow mode. It should be able to read the LVR of each attached i2c device to determine if it should be mixed fast or mixed slow (or pure if no i2c devices on the bus(

Reply:
I have added the LVR logic check, but I am still verifying whether the i2c-scl-hz property is specified in the DTS, as it needs to be checked first according to the documentation in i3c-controller.yaml